src/main/kotlin/me/partlysanestudios/partlysaneskies/data/skyblockdata/SkyblockDataManager.kt

```diff
@@ -0,0 +1,195 @@
+                    time
+            }.start()
+            lastAhUpdateTime = time
+        }
+    }
+
+    //    --------------------------- Skills ---------------------------
+    private var idToSkillMap = HashMap<String, SkyblockSkill>()
+    @Throws(MalformedURLException::class)
+    fun initSkills() {
+        newRequest(
+            Request("https://api.hypixel.net/resources/skyblock/skills",
+                RequestRunnable { s: Request ->
+                    val itemDataString = s.getResponse()
+                    if (!s.hasSucceeded()) {
+                        return@RequestRunnable
+                    }
+                    val skillDataJson =
+                        JsonParser().parse(itemDataString).getAsJsonObject()
+                    val skillObject = skillDataJson["skills"].getAsJsonObject()
+                    idToSkillMap =
+                        HashMap()
+
+//            Goes through each skill
+                    for ((id, value) in skillObject.entrySet()) {
+//                Gets the id from the key of the element
+                        //                Gets the data from the skill
+                        val `object` = value.getAsJsonObject()
+
+//                Gets the max level
+                        val maxLevel = `object`["maxLevel"].asInt
+                        //                Gets the json array containing the level
+                        val levelArray = `object`.getAsJsonArray("levels")
+                        //                Creates a new hashmap to store the level and the total Experience Required to get to that level
+                        val levelToExpMap =
+                            HashMap<Int, Float>()
+                        //                Adds the level to the map
+                        for (i in 0 until levelArray.size()) {
+                            val level = i + 1
+                            val experience =
+                                levelArray[i].getAsJsonObject()["totalExpRequired"].asFloat
+                            levelToExpMap[level] = experience
+                        }
+                        //                Instantiates a new skyblock skill with all the data
+                        val skill = SkyblockSkill(id, maxLevel, levelToExpMap)
+                        //                Adds the skill to the map
+                        idToSkillMap[id] =
+                            skill
+                    }
+                }, false, false
+            )
+        )
+    }
+
+    fun getSkill(skillId: String): SkyblockSkill? {
+        return idToSkillMap[skillId]
+    }
+
+    //    --------------------------- Players ---------------------------
+    private val playerCache = HashMap<String, SkyblockPlayer>()
+    fun getPlayer(username: String): SkyblockPlayer {
+        val player: SkyblockPlayer
+        if (playerCache.containsKey(username)) {
+            player = playerCache[username]!! // we can !! because we check above
```

Comment:
you wouldnt be able to pr into sh lmao
(no need to fix smth)

Reply:
We check above but fine I'll fix it

Reply:
noo no need to
just leave it as is